cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -12,113 +12,150 @@
 package processes
 
 import (
-	"bufio"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"os"
-	"path"
+	"path/filepath"
 	"strings"
 	"time"
 
+	"github.com/palantir/stacktrace"
 	"golang.org/x/crypto/ssh"
 
 	"alluxio.org/cli/env"
 	"alluxio.org/log"
 )
 
-func getMasters() ([]string, error) {
-	mastersDir := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioConfDir.EnvVar), "masters")
-	mastersFile, err := os.Open(mastersDir)
-	if err != nil {
-		log.Logger.Errorf("Error reading worker hostnames at %s", mastersDir)
-		return nil, err
+var cliPath = filepath.Join(env.Env.EnvVar.GetString(env.ConfAlluxioHome.EnvVar), "bin", "alluxio")
+var privateKeySigner ssh.Signer
+var masterList []string
+var workerList []string
+
+// for each master, create a client and run
+// TODO: now start master one by one, need to do them in parallel
+func runCommand(command string, onMasters bool, onWorkers bool) error {
+	// get list of masters and workers
+	var nodes []string
+	if onMasters {
+		if len(masterList) == 0 {
+			if err := getNodes(true); err != nil {
+				log.Logger.Fatalf("Cannot get masters, error: %s", err)
+			}
+		}
+		nodes = append(nodes, masterList...)
+	}
+	if onWorkers {
+		if len(workerList) == 0 {
+			if err := getNodes(false); err != nil {
+				log.Logger.Fatalf("Cannot get workers, error: %s", err)
+			}
+		}
+		nodes = append(nodes, workerList...)
+	}
+
+	// get public key if nones
+	if privateKeySigner == nil {
+		if err := getPrivateKey(); err != nil {
+			log.Logger.Fatalf("Cannot get private key, error: %s", err)
+		}
 	}
 
-	mastersReader := bufio.NewReader(mastersFile)
-	var mastersList []string
-	lastLine := false
-	for !lastLine {
-		// read lines of the workers file
-		line, err := mastersReader.ReadString('\n')
+	// dial nodes, run sessions and close
+	var errors []error
+	for _, node := range nodes {
+		conn, err := dialConnection(node, privateKeySigner)
 		if err != nil {
-			if err == io.EOF {
-				lastLine = true
-			} else {
-				log.Logger.Errorf("Error parsing worker file at this line: %s", line)
-				return nil, err
-			}
+			errors = append(errors, err)
+			continue
 		}
-		// remove notes
-		if strings.Index(line, "#") != -1 {
-			line = line[:strings.Index(line, "#")]
+		if err := runSession(node, conn, command); err != nil {
+			errors = append(errors, err)
 		}
-		line = strings.TrimSpace(line)
-		if line != "" {
-			mastersList = append(mastersList, line)
+		if err := conn.Close(); err != nil {
+			log.Logger.Infof("Connection to %s closed. Error: %s", node, err)
+		} else {
+			log.Logger.Infof("Connection to %s closed.", node)
 		}
 	}
-	return mastersList, nil
+
+	if len(errors) != 0 {
+		log.Logger.Warningf("Run command %s failed, number of failures: %v", command, len(errors))
+		return stacktrace.Propagate(errors[0], "First error: ")
```

Comment:
i'm not sure if we need to log every error here or dump it to some file for debugging. but i'm agree with the current structure: handling the error list, and return a digest error for the caller.